llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp

```diff
@@ -1340,6 +1340,13 @@ RISCVVLOptimizer::checkUsers(const MachineInstr &MI) const {
       continue;
     }
 
+    if (UserMI.isPHI()) {
+      LLVM_DEBUG(dbgs() << "    Peeking through uses of PHI\n");
+      for (auto &PhiUse : MRI->use_operands(UserMI.getOperand(0).getReg()))
+        Worklist.insert(&PhiUse);
```

Comment:
Do you have a test for a recurrence phi? I think locally I needed a "seen" set of pointers otherwise you ended up with infinite loops: https://github.com/lukel97/llvm-project/blob/a82b6466715ae03e5698a323cb2daae45dd2bded/llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp#L1331-L1353

Reply:
Oops I forgot to add context to that link. That branch is a WIP patch I have that generalizes the RISCVVLOptimizer pass to a dataflow analysis. I originally added it because I thought we needed to handle recurrences when removing vp intrinsics. But I didn't pursue it any further because it didn't have any impact on SPEC.

Reply:
Test added.